Improvements

alexkazik opened the pull request:
Add some small improvements.

alexkazik committed:
Improve open_dir

alexkazik committed:
Flush on unmount

alexkazik committed:
Expose bytes_per_sector and first_cluster

alexkazik committed:
Fix clippy warnings

rafalh commented on src/dir_entry.rs:
```diff
@@ -606,7 +606,9 @@ impl<'a, IO: ReadWriteSeek, TP, OCC: OemCpConverter> DirEntry<'a, IO, TP, OCC> {
-    pub(crate) fn first_cluster(&self) -> Option<u32> {
+    /// Return first cluster of a entry.
+    #[must_use]
+    pub fn first_cluster(&self) -> Option<u32> {
```
why make it public? How the user can use this information?

alexkazik replied:
I use it as a identifier (while a fs is mounted), since it's much simpler than filenames (due to lfn and only Option<u32> to be stored than a full filename) in my no_std, no_alloc project.

rafalh replied:
It's bad identifier, because empty file does not have it. You could truncate a file and write to it and again, this id would change without deleting the file.
Better identifier would be to use directory entry offset, but I'm not sure providing such id without any mean of using it (e.g. opening file by id) makes sense, and opening file by id is pretty much impossible without making it possible to corrupt the filesystem (see #109).
I propose to remove this change from this PR and if you still need it a separate PR would be better.

alexkazik replied:
I don't intend to create a find/open by id function.
In my case it's just used to remember (with limited storage and without alloc) the file.
Maybe add a warning or make it unsafe - though it's not doing unsafe stuff, so that may be misleading.

rafalh replied:
I still don't understand how exactly do you use this information. Is it some kind of visited files list and you use it to detect new files or something like that?
Why wouldn't my idea of using directory entry offset work? It seems more reliable and allows IDs for empty files.
This should probably be a separate PR and not part of a generic one titled "Improvements". Also tests for new api would be welcome.

rafalh commented on src/dir.rs:
```diff
@@ -200,20 +200,33 @@ impl<'a, IO: ReadWriteSeek, TP: TimeProvider, OCC: OemCpConverter> Dir<'a, IO, T
     ///
     /// `path` is a '/' separated directory path relative to self directory.
     ///
+    /// An empty path returns the current directory.
```
I'm not sure this is a good way to handle this situation. Do you know any other API that behaves like this? What problem does this change solve?

rafalh replied:
Yes, I want an error in case of an empty string, because that's not how this function is supposed to be called. User can always use `.` if they want to open the current directory (but `dir.clone()` would be simpler).
The new code is much longer which I don't like, but if you remove the `if` it will look a little better. `self.clone()` is a little worrying, but it's only copying some data and doesn't cause dynamic allocation. The new code uses iteration instead of recursion which is good, because it's easier to reason about stack memory usage, so I guess I can accept it if you remove empty string special handling

rafalh commented on src/dir.rs:
```diff
@@ -200,20 +200,33 @@ impl<'a, IO: ReadWriteSeek, TP: TimeProvider, OCC: OemCpConverter> Dir<'a, IO, T
+                return Ok(dir);
+            }
+            let (name, rest_opt) = split_path(path);
+            let e = dir.find_entry(name, Some(true), None)?.to_dir();
```
Name `e` no longer make sense here. Please remove it and update `dir` instead

alexkazik replied:
updated

rafalh reviewed (changes requested):
I like some of the improvements here, but not all. I would prefer to split controversial changes into a separate PR. Also it would be nice if changelog file was also updated

alexkazik commented:
> I like some of the improvements here, but not all. I would prefer to split controversial changes into a separate PR. Also it would be nice if changelog file was also updated

If you tell me what is controversial, I'll remove it (and maybe add in another PR). (I guess it's the directory routine.)

Added to changelog.

(P.S. I'd also appreciate it if PR https://github.com/rafalh/rust-fatfs/pull/78 and https://github.com/rafalh/rust-fatfs/pull/98 can also be merged.)

rafalh commented on CHANGELOG.md:
```diff
@@ -41,13 +41,16 @@ New features:
 * Undeprecate `set_created`, `set_accessed`, `set_modified` methods on `File` - there are scenarios where those methods are needed
 * Relax BPB validation by allowing non-zero values in both `total_sectors_32` and `total_sectors_16`
 * Add `strict` field in `FsOptions`, which allows disabling validation of boot signature to improve compatibility with old FAT images
+* Improve `open_dir`
```
"Improve" is not really helpful. I guess you could write that you changed path parsing from recursion to iteration

rafalh commented on CHANGELOG.md:
```diff
@@ -41,13 +41,16 @@ New features:
 * Relax BPB validation by allowing non-zero values in both `total_sectors_32` and `total_sectors_16`
 * Add `strict` field in `FsOptions`, which allows disabling validation of boot signature to improve compatibility with old FAT images
+* Improve `open_dir`
+* Expose bytes_per_sector and first_cluster
```
If you write function names please use `code` formatting

rafalh reviewed (commented):
I don't know if you still want to work on this PR as it took me a little long to review (sorry). If not I will probably integrate some of those changes myself

alexkazik commented:
I've updated the PR.

rafalh commented on src/dir.rs:
```diff
@@ -200,20 +200,33 @@ impl<'a, IO: ReadWriteSeek, TP: TimeProvider, OCC: OemCpConverter> Dir<'a, IO, T
+        }
+        let mut dir = self.clone();
+        loop {
+            if path.is_empty() {
```
I through you will delete both ifs (L214 and L219) after I asked to not support empty path. It seems you want to not only change from recursive style to iterative, but also make some changes to the behavior, but I'm not sure why. 
`split_path` already handles the trimming of `/` including trailing separator. It seems impossible for the second if from line 219 to be ever activated. If path "foo/" is passed to `split_path` it would return `("foo", None)` and loop will end. The `if` from line 214 also seems redundant, because `find_entry` with empty string will return an error.
If there is API change needed it would be best to put it in a separate PR and add a test for it, so new behavior won't be lost in some future refactors.

rafalh commented on CHANGELOG.md:
```diff
@@ -41,13 +41,16 @@ New features:
 * Fill FAT32 root directory clusters with zeros after allocation to avoid interpreting old data as directory entries
 * Put '.' and '..' in the first two directory entries. (fixes "Expected a valid '.' entry in this slot." fsck error)
 * Set the cluster number to 0 in the ".." directory entry if it points to the root dir
+* Flush on unmount
```
To be honest I don't think not explicitly flushing in `unmount` is a bug. `unmount` drops `self` and in consequence drops the disk, and I would expect the disk object to write all changes when it is dropped. 
Putting this change in bug fixes without any explanation suggests that previous versions could cause corruption by not writing some changes, which I don't think is likely. The only change it that now `unmount` will return an error if writing the changes (flushing) fails.
I think it could be put in Improvements section instead
